[Site Isolation] Owner element should be preserved when switching between LocalFrame and RemoteFrame

achristensen07 opened the pull request:
#### 6f62e90b5cac93d820552a99e2241e9a90875514
<pre>
[Site Isolation] Owner element should be preserved when switching between LocalFrame and RemoteFrame
<a href="https://bugs.webkit.org/show_bug.cgi?id=273809">https://bugs.webkit.org/show_bug.cgi?id=273809</a>
<a href="https://rdar.apple.com/127638179">rdar://127638179</a>

Reviewed by Charlie Wolfe.

Otherwise it suddenly becomes null from the point of view of the Frame and DOMWindow,
which prevents things like the load event from flowing to the event listeners.

I also rename createSubframeHostedInAnotherProcess to createProvisionalSubframe because
the contents aren&apos;t necessarily hosted in another process.

This test hit a case where sometimes a frame was created and destroyed before the frame
we transitioned from was destroyed, so I updated the assertion to not assert in that case.

* LayoutTests/http/tests/site-isolation/load-event-after-transition-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/load-event-after-transition.html: Added.
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::setOwnerElement):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::createProvisionalSubframe):
(WebCore::LocalFrame::createSubframeHostedInAnotherProcess): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createProvisionalFrame):
(WebKit::WebFrame::commitProvisionalFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:

Canonical link: <a href="https://commits.webkit.org/278472@main">https://commits.webkit.org/278472@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdaf106227c21b796f0bd5a7abcada4bdb57f4fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2957 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53899 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1331 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36197 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/981 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43620 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22401 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24982 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/876 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9081 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46961 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/938 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55489 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25742 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/841 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48698 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43764 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47753 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27867 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7336 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26731 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -387,6 +387,7 @@ void WebFrame::loadDidCommitInAnotherProcess(std::optional<WebCore::LayerHosting
         frame->setOpener(newFrame.ptr());
 
     newFrame->tree().setSpecifiedName(localFrame->tree().specifiedName());
+    newFrame->setOwnerElement(ownerElement.get());
```
We already pass `ownerElement` into `RemoteFrame::createSubframeWithContentsInAnotherProcess`. We could also pass it into `RemoteFrame::createSubframe` and avoid this call. Fine either way though.

charliewolfe replied:
Or we could remove the ownerElement argument from `createSubframeWithContentsInAnotherProcess`.

achristensen07 replied:
And the test could be a layout test.

achristensen07 replied:
Yeah, this call to setOwnerElement is not needed but the other is.  We can't pass it in to the constructor because the owner element shouldn't know about a provisional frame until it commits.

achristensen07 committed:
[Site Isolation] Owner element should be preserved when switching between LocalFrame and RemoteFrame
https://bugs.webkit.org/show_bug.cgi?id=273809
rdar://127638179

Reviewed by Charlie Wolfe.

Otherwise it suddenly becomes null from the point of view of the Frame and DOMWindow,
which prevents things like the load event from flowing to the event listeners.

I also rename createSubframeHostedInAnotherProcess to createProvisionalSubframe because
the contents aren't necessarily hosted in another process.

This test hit a case where sometimes a frame was created and destroyed before the frame
we transitioned from was destroyed, so I updated the assertion to not assert in that case.

* LayoutTests/http/tests/site-isolation/load-event-after-transition-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/load-event-after-transition.html: Added.
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::setOwnerElement):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::createProvisionalSubframe):
(WebCore::LocalFrame::createSubframeHostedInAnotherProcess): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createProvisionalFrame):
(WebKit::WebFrame::commitProvisionalFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:

Canonical link: https://commits.webkit.org/278472@main
